File: roms/golden_axe.py
#
#	Golden Axe
#

from base64 import b64encode
from io import BytesIO
from math import ceil
from sys import argv
from zipfile import ZipFile
from PIL import Image

with ZipFile(argv[1]) as z:
    merge = lambda even, odd : bytes([odd[i // 2] if i % 2 else even[i // 2] for i in range(len(even) * 2)])
    prg1 = merge(z.read('goldnaxej/epr-12540.a7') + z.read('goldnaxe2/epr-12521.a8'), z.read('goldnaxej/epr-12539.a5') + z.read('goldnaxe2/epr-12519.a6'))
    key = z.read('goldnaxej/317-0121.key')
    bg = z.read('epr-12385.ic19') + z.read('epr-12386.ic20') + z.read('epr-12387.ic21')
    obj = merge(z.read('mpr-12379.ic12')[:0x20000] + z.read('mpr-12381.ic13')[:0x20000], z.read('mpr-12378.ic9')[:0x20000] + z.read('mpr-12380.ic10')[:0x20000])
    obj += merge(z.read('mpr-12383.ic14')[:0x20000], z.read('mpr-12382.ic11')[:0x20000]) + b'\xff' * 0x40000
    obj += merge(z.read('mpr-12379.ic12')[0x20000:] + z.read('mpr-12381.ic13')[0x20000:], z.read('mpr-12378.ic9')[0x20000:] + z.read('mpr-12380.ic10')[0x20000:])
    obj += merge(z.read('mpr-12383.ic14')[0x20000:], z.read('mpr-12382.ic11')[0x20000:]) + b'\xff' * 0x40000
    prg2 = z.read('epr-12390.ic8') + z.read('mpr-12384.ic6')

rom = prg1 + key + bg + obj + prg2

def pngstring(a):
    w = 1024
    img = Image.new('P', (w, ceil(len(a) / w)))
    img.putpalette(sum([[i, 0, 0] for i in range(256)], []))
    img.putdata(a)
    buf = BytesIO()
    img.save(buf, 'PNG')
    return b64encode(buf.getvalue())

with open(argv[2], 'wb') as f:
    f.write(b'export const ROM = \'data:image/png;base64,' + pngstring(rom) + b'\';\n')

File: roms/golden_axe.js
/*
 *	Golden Axe
 */

const fs = require('fs');
const {createCanvas, createImageData} = require('canvas');
const {Zlib} = require('zlibjs/bin/unzip.min');
Zlib.Unzip.prototype.read = Zlib.Unzip.prototype.decompress;

const cat = (...args) => {
	const array = new Uint8Array(args.reduce((a, b) => a + b.length, 0));
	return args.reduce((a, b) => (array.set(b, a), a + b.length), 0), array;
};
const merge = (...args) => {
	const array = new Uint8Array(args.reduce((a, b) => a + b.length, 0)), step = args.length;
	return args.forEach((a, i) => a.forEach((e, j) => array[j * step + i] = e)), array;
};
const pad = (length, value = 0xff) => new Uint8Array(length).fill(value);

const z = new Zlib.Unzip(fs.readFileSync(process.argv[2]));
const prg1 = merge(cat(z.read('goldnaxej/epr-12540.a7'), z.read('goldnaxe2/epr-12521.a8')), cat(z.read('goldnaxej/epr-12539.a5'), z.read('goldnaxe2/epr-12519.a6')));
const key = z.read('goldnaxej/317-0121.key');
const bg = cat(z.read('epr-12385.ic19'), z.read('epr-12386.ic20'), z.read('epr-12387.ic21'));
let obj = merge(cat(z.read('mpr-12379.ic12').subarray(0, 0x20000), z.read('mpr-12381.ic13').subarray(0, 0x20000)), cat(z.read('mpr-12378.ic9').subarray(0, 0x20000), z.read('mpr-12380.ic10').subarray(0, 0x20000)));
obj = cat(obj, merge(z.read('mpr-12383.ic14').subarray(0, 0x20000), z.read('mpr-12382.ic11').subarray(0,0x20000)), pad(0x40000));
obj = cat(obj, merge(cat(z.read('mpr-12379.ic12').subarray(0x20000), z.read('mpr-12381.ic13').subarray(0x20000)), cat(z.read('mpr-12378.ic9').subarray(0x20000), z.read('mpr-12380.ic10').subarray(0x20000))));
obj = cat(obj, merge(z.read('mpr-12383.ic14').subarray(0x20000), z.read('mpr-12382.ic11').subarray(0x20000)), pad(0x40000));
const prg2 = cat(z.read('epr-12390.ic8'), z.read('mpr-12384.ic6'));

const rom = cat(prg1, key, bg, obj, prg2);

const pngString = array => {
	const w = 1024, h = Math.ceil(array.byteLength / w), canvas = createCanvas(w, h);
	const palette = new Uint8ClampedArray(new Uint32Array(256).map((e, i) => i | 0xff000000).buffer);
	canvas.getContext('2d', {pixelFormat: 'A8'}).putImageData(createImageData(new Uint8ClampedArray(array.buffer), w, h), 0, 0);
	return Buffer.from(canvas.toBuffer('image/png', {compressionLevel: 9, filters: canvas.PNG_FILTER_NONE, palette})).toString('base64');
};

fs.writeFileSync(process.argv[3], `export const ROM = 'data:image/png;base64,${pngString(rom)}';\n`);
